test/e2e/framework/job/wait.go

```diff
@@ -82,14 +83,17 @@ func waitForJobPodsInPhase(ctx context.Context, c clientset.Interface, ns, jobNa
 // both conformance CI jobs with GA-only features and e2e CI jobs with all default-enabled features.
 // So, we need to skip "Complete" condition reason verifications in the e2e conformance test cases.
 func WaitForJobComplete(ctx context.Context, c clientset.Interface, ns, jobName string, reason *string, completions int32) error {
-	if err := wait.PollUntilContextTimeout(ctx, framework.Poll, JobTimeout, false, func(ctx context.Context) (bool, error) {
-		curr, err := c.BatchV1().Jobs(ns).Get(ctx, jobName, metav1.GetOptions{})
+	if err := framework.Gomega().Eventually(ctx, func() int32 {
+		job, err := c.BatchV1().Jobs(ns).Get(ctx, jobName, metav1.GetOptions{})
 		if err != nil {
```

Comment:
I suppose, the best would be to allow for only a couple of failures. @pohly thoughts?

Reply:
That's exactly the distinction that the test author needs to make: is an error retryable or permanent.

The `framework.HandleRetry` handler has some logic for API server errors. If you don't want to use it, then better treat all such errors as retryable and only use `StopPolling` for conditions which are definitely recognized as "this will definitely not change anymore".

Reply:
+1 on retrying the API failures. I think the previous approach would fail on API server errors, so I thought it could be a follow up PR. OTOH it is so closely related that we may as well use `HandleRetry` in this PR.

Reply:
Makes sense, so in this case it makes sense to go back to the previous implementation (which can maybe be refactored to be less verbose) and use `HandleRetry`. WDYT?

Reply:
sgtm, let's try to use HandleRetry

Reply:
@mimowo @pohly I've implemented the solution using `HandleRetry`. Now the `get` function returns a pointer to the job instead of a bool. In that way, in case of a non-transient error, `HandleRetry` [produces a meaningful message](https://github.com/kubernetes/kubernetes/blob/7ab8a2b92f675bb81715bfca7085b0c4d0a284c9/test/e2e/framework/get.go#L88) apart from stop polling. I've tested all these situations (fail fast, non-transient errors.. etc) and seems to work well.

P.S: @pohly out of curiosity, I've tested the same situations with the previous implementation and also seemed to work well 🤷🏻